WebDriverAgentLib/FBSpringboardApplication.m

```diff
@@ -36,11 +38,37 @@ + (instancetype)fb_springboard
 
 - (BOOL)fb_tapApplicationWithIdentifier:(NSString *)identifier error:(NSError **)error
 {
-  XCUIElement *appElement = [[self descendantsMatchingType:XCUIElementTypeAny]
-                             elementMatchingPredicate:[NSPredicate predicateWithFormat:@"%K = %@", FBStringify(XCUIElement, identifier), identifier]
-                             ];
-  if (![appElement fb_scrollToVisibleWithNormalizedScrollDistance:1.0 scrollDirection:FBXCUIElementScrollDirectionHorizontal error:error]) {
-    return NO;
+  XCUIElementQuery *appElementsQuery = [[self descendantsMatchingType:XCUIElementTypeIcon] matchingIdentifier:identifier];
+  NSArray<XCUIElement *> *matchedAppElements = [appElementsQuery allElementsBoundByIndex];
+  if (0 == matchedAppElements.count) {
+    return [[[FBErrorBuilder builder]
+      withDescriptionFormat:@"Cannot locate Springboard icon for '%@' application", identifier]
+     buildError:error];
+  }
+  // Select the most recent installed application if there are multiple matches
+  XCUIElement *appElement = [matchedAppElements lastObject];
+  if (!appElement.fb_isVisible) {
+    CGRect startFrame = appElement.frame;
+    BOOL shouldSwipeToTheRight = startFrame.origin.x < 0;
+    NSString *errorDescription = [NSString stringWithFormat:@"Cannot scroll to Springboard icon for '%@' application", identifier];
+    do {
```

Comment:
because there is a possibility (a very little one, but still) that swipe didn't work and thus the icon will never be visible. this can lead to an endless loop and freeze %)